pkg/reconciler/timeout_handler_test.go

```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
 	testHandler.CheckTimeouts()
 
 }
+
+// TestGetTimeout checks that the timeout calculated for a given taskrun falls
+// back to the default taskrun timeout when none is provided.
+func TestGetTimeout(t *testing.T) {
+	testCases := []struct {
+		description      string
+		inputDuration    *metav1.Duration
+		expectedDuration time.Duration
+	}{
+		{
+			description:      "returns same duration as input when input is not nil",
+			inputDuration:    &metav1.Duration{2 * time.Second},
+			expectedDuration: 2 * time.Second,
+		},
+		{
+			description:      "returns an end time using the default timeout if a TaskRun's timeout is nil",
+			inputDuration:    nil,
+			expectedDuration: defaultTimeout,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			receivedDuration := GetTimeout(tc.inputDuration)
+			if receivedDuration != tc.expectedDuration {
+				t.Errorf("expected %q received %q", tc.expectedDuration.String(), receivedDuration.String())
+			}
+		})
+	}
+}
+
+// TestSetTaskRunTimer checks that the SetTaskRunTimer method correctly calls the TaskRun
+// callback after a set amount of time.
+func TestSetTaskRunTimer(t *testing.T) {
+	taskRun := tb.TaskRun("test-taskrun-arbitrary-timer", testNs, tb.TaskRunSpec(
+		tb.TaskRunTaskRef(simpleTask.Name, tb.TaskRefAPIVersion("a1")),
+		tb.TaskRunTimeout(2*time.Second),
+	), tb.TaskRunStatus(tb.Condition(apis.Condition{
+		Type:   apis.ConditionSucceeded,
+		Status: corev1.ConditionUnknown}),
+		tb.TaskRunStartTime(time.Now().Add(-10*time.Second)),
+	))
+
+	d := test.Data{
+		TaskRuns: []*v1alpha1.TaskRun{taskRun},
+		Tasks:    []*v1alpha1.Task{simpleTask},
+		Namespaces: []*corev1.Namespace{{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testNs,
+			},
+		}},
+	}
+	stopCh := make(chan struct{})
+	c, _ := test.SeedTestData(t, d)
+	observer, _ := observer.New(zap.InfoLevel)
+	testHandler := NewTimeoutHandler(c.Kube, c.Pipeline, stopCh, zap.New(observer).Sugar())
+	timerDuration := 50 * time.Millisecond
+	timerFailDeadline := 100 * time.Millisecond
+	doneCh := make(chan struct{})
+	callback := func(_ interface{}) {
+		close(doneCh)
```

Comment:
you could also have this function set a boolean or something that would could use to assert that the function was called or not called as expected

Reply:
I initially implemented it using a bool that the callback mutated but this felt a little icky due to the inline sleep:

```go
done := false
testHandler.SetTaskRunCallbackFunc(func(_ interface{}) { done = true })
go testHandler.SetTaskRunTimer(taskRun, timerDuration)
<-time.After(timerFailDeadline) // ick?
if !done {
  t.Errorf("timer did not execute task run callback func within expected time")
}
```

I kind of like how the `select` approach races the two channels against each other.  Full credit to @dlorenc (https://github.com/tektoncd/pipeline/pull/905#discussion_r287564205) for pointing this out.

I may have completely misinterpreted either or both approaches though so lmk if there's something I'm missing.